craft_parts/executor/executor.py

```diff
@@ -0,0 +1,199 @@
+        extra_build_snaps: List[str] = None,
+    ):
+        self._part_list = part_list
+        self._project_info = project_info
+        self._extra_build_packages = extra_build_packages
+        self._extra_build_snaps = extra_build_snaps
+        self._handler: Dict[str, PartHandler] = {}
+
+    def prologue(self):
+        """Prepare the execution environment.
+
+        This method is called before executing lifecycle actions.
+        """
+        self._install_build_packages()
+        self._install_build_snaps()
+
+        callbacks.run_prologue(self._project_info, part_list=self._part_list)
+
+    def epilogue(self):
+        """Finish and clean the execution environment.
+
+        This method is called after executing lifecycle actions.
+        """
+        callbacks.run_epilogue(self._project_info, part_list=self._part_list)
+
+    def execute(self, actions: Union[Action, List[Action]]) -> None:
+        """Execute the specified action or list of actions.
+
+        :param actions: An :class:`Action` object or list of :class:`Action`
+           objects specifying steps to execute.
+
+        :raises InvalidActionException: If the action parameters are invalid.
+        """
+        if isinstance(actions, Action):
+            actions = [actions]
+
+        for act in actions:
+            self._run_action(act)
+
+    def clean(self, initial_step: Step, *, part_names: List[str] = None) -> None:
+        """Clean the given parts, or all parts if none is specified.
+
+        :param initial_step: The step to clean. More steps may be cleaned
+            as a consequence of cleaning the initial step.
+        :param part_names: A list with names of the parts to clean. If not
+            specified, all parts will be cleaned.
+        """
+        selected_parts = parts.part_list_by_name(part_names, self._part_list)
+
+        selected_steps = [initial_step] + initial_step.next_steps()
+        selected_steps.reverse()
+
+        for part in selected_parts:
+            self._create_part_handler(part)
+            handler = self._handler[part.name]
+
+            for step in selected_steps:
+                handler.clean_step(step=step)
+
+    def _run_action(self, action: Action):
+        """Execute the given action for a part using the provided step information.
+
+        :param action: The lifecycle action to run.
+        """
+        part = parts.part_by_name(action.part_name, self._part_list)
+
+        logger.debug("execute action %s:%s", part.name, action)
+
+        if action.action_type == ActionType.SKIP:
+            logger.debug("Skip execution of %s (because %s)", action, action.reason)
+            return
+
+        if action.step == Step.STAGE:
+            check_for_stage_collisions(self._part_list)
+
+        self._create_part_handler(part)
+
+        handler = self._handler[part.name]
+        handler.run_action(action)
+
+    def _create_part_handler(self, part: Part):
```

Comment:
```suggestion
    def _create_part_handler(self, part: Part) -> None:
```

But may be nicer if also returning the part.